File: java/src/main/java/io/webweavex/determinism/PyJson.java
package io.webweavex.determinism;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Faithful Python {@code json.dumps} for the native value tree — configurable
 * separators and {@code ensure_ascii}, always {@code sort_keys=True}.
 *
 * <p>Unlike {@link CanonicalJson} this does <em>not</em> canonicalize numbers or
 * strip volatile keys; it reproduces {@code json.dumps} verbatim, which several
 * canonical runtimes hash directly (e.g. {@code compute_global_runtime_fingerprint}
 * uses {@code separators=(",",":")} and the default {@code ensure_ascii=True};
 * {@code _graph_hash} uses the default {@code (", ", ": ")} separators).
 */
public final class PyJson {

    private PyJson() {
    }

    /** {@code json.dumps(v, sort_keys=True, separators=(",",":"))} (ensure_ascii default True). */
    public static String dumpsCompactAscii(Object v) {
        return dumps(v, ",", ":", true);
    }

    /** {@code json.dumps(v, sort_keys=True)} — default separators {@code (", ", ": ")}, ascii. */
    public static String dumpsDefaultAscii(Object v) {
        return dumps(v, ", ", ": ", true);
    }

    /** {@code json.dumps(v, ensure_ascii=False, sort_keys=True, separators=(",",":"))}. */
    public static String dumpsCompactUnicode(Object v) {
        return dumps(v, ",", ":", false);
    }

    public static String dumps(Object value, String itemSep, String kvSep, boolean ensureAscii) {
        StringBuilder sb = new StringBuilder();
        encode(value, sb, itemSep, kvSep, ensureAscii);
        return sb.toString();
    }

    private static void encode(Object v, StringBuilder sb, String itemSep, String kvSep,
                               boolean ensureAscii) {
        if (v == null) {
            sb.append("null");
        } else if (v instanceof Boolean) {
            sb.append(((Boolean) v) ? "true" : "false");
        } else if (v instanceof Double || v instanceof Float) {
            sb.append(floatRepr(((Number) v).doubleValue()));
        } else if (v instanceof Number) {
            sb.append(v.toString());
        } else if (v instanceof String) {
            escape((String) v, sb, ensureAscii);
        } else if (v instanceof List) {
            List<?> list = (List<?>) v;
            sb.append('[');
            for (int i = 0; i < list.size(); i++) {
                if (i > 0) {
                    sb.append(itemSep);
                }
                encode(list.get(i), sb, itemSep, kvSep, ensureAscii);
            }
            sb.append(']');
        } else if (v instanceof Map) {
            Map<?, ?> map = (Map<?, ?>) v;
            List<String> keys = new ArrayList<>(map.size());
            for (Object k : map.keySet()) {
                keys.add(String.valueOf(k));
            }
            keys.sort(Normalization::codePointCompare);
            sb.append('{');
            boolean first = true;
            for (String k : keys) {
                if (!first) {
                    sb.append(itemSep);
                }
                first = false;
                escape(k, sb, ensureAscii);
                sb.append(kvSep);
                encode(lookup(map, k), sb, itemSep, kvSep, ensureAscii);
            }
            sb.append('}');
        } else {
            escape(v.toString(), sb, ensureAscii);
        }
    }

    /** Python {@code json} float form: {@code repr} for finite, {@code NaN}/{@code Infinity} otherwise. */
    private static String floatRepr(double d) {
        if (Double.isNaN(d)) {
            return "NaN";
        }
        if (d == Double.POSITIVE_INFINITY) {
            return "Infinity";
        }
        if (d == Double.NEGATIVE_INFINITY) {
            return "-Infinity";
        }
        return PyFloat.pyFloatRepr(d);
    }

    private static Object lookup(Map<?, ?> map, String key) {
        if (map.containsKey(key)) {
            return map.get(key);
        }
        for (Map.Entry<?, ?> e : map.entrySet()) {
            if (String.valueOf(e.getKey()).equals(key)) {
                return e.getValue();
            }
        }
        return null;
    }

    private static void escape(String s, StringBuilder sb, boolean ensureAscii) {
        sb.append('"');
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            switch (c) {
                case '"':
                    sb.append("\\\"");
                    break;
                case '\\':
                    sb.append("\\\\");
                    break;
                case '\b':
                    sb.append("\\b");
                    break;
                case '\t':
                    sb.append("\\t");
                    break;
                case '\n':
                    sb.append("\\n");
                    break;
                case '\f':
                    sb.append("\\f");
                    break;
                case '\r':
                    sb.append("\\r");
                    break;
                default:
                    if (c < 0x20 || (ensureAscii && c > 0x7E)) {
                        sb.append(String.format("\\u%04x", (int) c));
                    } else {
                        sb.append(c);
                    }
            }
        }
        sb.append('"');
    }
}

File: kotlin/src/main/kotlin/io/webweavex/determinism/PyJson.kt
package io.webweavex.determinism

/**
 * Faithful Python `json.dumps` for the native value tree — configurable
 * separators and `ensure_ascii`, always `sort_keys=True`.
 *
 * Unlike [CanonicalJson] this does *not* canonicalize numbers or strip
 * volatile keys; it reproduces `json.dumps` verbatim. Mirrors
 * `io.webweavex.determinism.PyJson` (Java) line for line.
 */
object PyJson {

    /** `json.dumps(v, sort_keys=True, separators=(",",":"))` (ensure_ascii default True). */
    @JvmStatic
    fun dumpsCompactAscii(v: Any?): String = dumps(v, ",", ":", true)

    /** `json.dumps(v, sort_keys=True)` — default separators `(", ", ": ")`, ascii. */
    @JvmStatic
    fun dumpsDefaultAscii(v: Any?): String = dumps(v, ", ", ": ", true)

    /** `json.dumps(v, ensure_ascii=False, sort_keys=True, separators=(",",":"))`. */
    @JvmStatic
    fun dumpsCompactUnicode(v: Any?): String = dumps(v, ",", ":", false)

    @JvmStatic
    fun dumps(value: Any?, itemSep: String, kvSep: String, ensureAscii: Boolean): String {
        val sb = StringBuilder()
        encode(value, sb, itemSep, kvSep, ensureAscii)
        return sb.toString()
    }

    private fun encode(v: Any?, sb: StringBuilder, itemSep: String, kvSep: String, ensureAscii: Boolean) {
        when (v) {
            null -> sb.append("null")
            is Boolean -> sb.append(if (v) "true" else "false")
            is Double -> sb.append(floatRepr(v))
            is Float -> sb.append(floatRepr(v.toDouble()))
            is Number -> sb.append(v.toString())
            is String -> escape(v, sb, ensureAscii)
            is List<*> -> {
                sb.append('[')
                for ((i, item) in v.withIndex()) {
                    if (i > 0) sb.append(itemSep)
                    encode(item, sb, itemSep, kvSep, ensureAscii)
                }
                sb.append(']')
            }
            is Map<*, *> -> {
                val keys = v.keys.map { it.toString() }.sortedWith { a, b -> Normalization.codePointCompare(a, b) }
                sb.append('{')
                var first = true
                for (k in keys) {
                    if (!first) sb.append(itemSep)
                    first = false
                    escape(k, sb, ensureAscii)
                    sb.append(kvSep)
                    encode(lookup(v, k), sb, itemSep, kvSep, ensureAscii)
                }
                sb.append('}')
            }
            else -> escape(v.toString(), sb, ensureAscii)
        }
    }

    /** Python `json` float form: `repr` for finite, `NaN`/`Infinity` otherwise. */
    private fun floatRepr(d: Double): String {
        if (d.isNaN()) return "NaN"
        if (d == Double.POSITIVE_INFINITY) return "Infinity"
        if (d == Double.NEGATIVE_INFINITY) return "-Infinity"
        return PyFloat.pyFloatRepr(d)
    }

    private fun lookup(map: Map<*, *>, key: String): Any? {
        if (map.containsKey(key)) return map[key]
        for ((k, v) in map) {
            if (k.toString() == key) return v
        }
        return null
    }

    private fun escape(s: String, sb: StringBuilder, ensureAscii: Boolean) {
        sb.append('"')
        for (c in s) {
            when (c) {
                '"' -> sb.append("\\\"")
                '\\' -> sb.append("\\\\")
                '\b' -> sb.append("\\b")
                '\t' -> sb.append("\\t")
                '\n' -> sb.append("\\n")
                '' -> sb.append("\\f")
                '\r' -> sb.append("\\r")
                else -> {
                    if (c.code < 0x20 || (ensureAscii && c.code > 0x7E)) {
                        sb.append(String.format("\\u%04x", c.code))
                    } else {
                        sb.append(c)
                    }
                }
            }
        }
        sb.append('"')
    }
}
